### PR TITLE
Build system fixes & improvements

### PR DESCRIPTION
**Pull request by johnnovak**
# Description

None of us are really CMake experts in the team, so as an experiment I asked Claude to review our entire CMake build system and offer improvement suggestions, fix bugs, and flag places where we're not doing things the idiomatic CMake way.

I think it did a very good job, I took basically 90% of the suggestions as I agree with the justifications (see the commit messages for details).

## Related issues

Depends on https://github.com/dosbox-staging/dosbox-staging/pull/4867

# Manual testing

CI is green.

The change has been manually tested on:

- [ ] Windows
- [x] macOS
- [ ] Linux

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/docs/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/docs/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/tools/compile-commits.sh) that all my commits can be built.
- [ ] my change has been manually tested on Windows, macOS, and Linux.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] provided the release notes draft (for significant user-facing changes).
- [x] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/docs/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/docs/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

**Review by kklobe — APPROVED** [review]
@johnnovak looks good on macOS for debug-macos and release-macos configs, tested with debugger enabled, everything worked there.

Maybe wait for a Linux confirmation before merging?

**Comment by johnnovak** [comment]
> @johnnovak looks good on macOS for debug-macos and release-macos configs, tested with debugger enabled, everything worked there.
> 
> Maybe wait for a Linux confirmation before merging?

Sure. @weirddan455 would you be able to test?

FYI, I've tested every single commit on Windows in Visual Studio — all build.

File: src/libs/glad/CMakeLists.txt [hunk 1]
```diff
@@ -1,13 +1,25 @@
...
-  target_include_directories(libglad PUBLIC release/include)
+add_library(libglad INTERFACE)
+add_dependencies(libglad libglad_debug libglad_release)
+target_link_libraries(libglad INTERFACE
+  $<IF:$<CONFIG:Debug>,libglad_debug,libglad_release>
+)
+target_include_directories(libglad INTERFACE
+  $<IF:$<CONFIG:Debug>,
+    ${CMAKE_CURRENT_SOURCE_DIR}/debug/include,
+    ${CMAKE_CURRENT_SOURCE_DIR}/release/include>
```
Comment (weirddan455) [inline src/libs/glad/CMakeLists.txt]: > The glad OpenGL loader has debug and release variants. The previous
approach used CMAKE_BUILD_TYPE to select between them, but this
variable is empty at configure time on multi-config generators
(Visual Studio, Xcode), so the debug variant was never used on
those platforms.

Did you confirm that this was actually a problem or is this just what the LLM told you? I don't see how this can be true since all of our presets explicitly set `CMAKE_BUILD_TYPE`.

This interface library song and dance is also harder to read and reason about. The old code had a simple branch. Now you've got 3 different libraries and 2 of these `IF CONFIG:DEBUG` branches plus more duplicated code in the Apple branch below.

File: CMakeLists.txt [hunk 2]
```diff
@@ -579,6 +579,7 @@ target_link_libraries(dosbox PRIVATE
   libdosboxcommon
   $<TARGET_NAME_IF_EXISTS:SDL2::SDL2main>
   $<IF:$<TARGET_EXISTS:SDL2::SDL2>,SDL2::SDL2,SDL2::SDL2-static>
+  $<$<BOOL:${SUPPORT_XINPUT2}>:${X11_Xinput_LIB}>
```
Comment (weirddan455) [inline CMakeLists.txt]: You left out an `IF` from the original. This is Linux specific but I don't know how to test this. `OPT_XINPUT` looks to be off by default and I only have one mouse. @FeralChild64 can you take a look at this?

File: CMakeLists.txt [hunk 3]
```diff
@@ -363,8 +363,8 @@ if (OPT_MANYMOUSE)
     message(WARNING "ManyMouse requires FD_ZERO support")
 
   elseif (DOSBOX_PLATFORM_MACOS)
-    include (${CMAKE_CURRENT_SOURCE_DIR}/src/libs/manymouse/cmake/FindIOKit.cmake)
-    if (NOT ${IOKit_FOUND})
+    find_library(IOKit_LIBRARY IOKit)
+    if (NOT IOKit_LIBRARY)
```
Comment (weirddan455) [inline CMakeLists.txt:372]: Again with the ManyMouse stuff that I have never used. Did you test this change?

File: src/libs/glad/CMakeLists.txt [hunk 4]
```diff
@@ -3,17 +3,17 @@
 # Multi-config generators (Xcode, Visual Studio) don't support per-config
 # source files, so we build both variants and select via INTERFACE wrapper.
 add_library(libglad_debug STATIC debug/src/gl.c)
-target_include_directories(libglad_debug PUBLIC debug/include)
+target_include_directories(libglad_debug SYSTEM PUBLIC debug/include)
```
Comment (weirddan455) [inline src/libs/glad/CMakeLists.txt]: > Adding SYSTEM tells the compiler to treat these paths as external
system headers, suppressing warnings from code we don't control.

This logic does not make sense to me. These are not system headers. They're in our git tree. Warnings should behave in the same way as the rest of this library `(debug/src/gl.c)` in this case.

If you want to suppress warnings in 3rd party libraries, that should be done explicitly. Marking an include directory as `SYSTEM` does more than only affect warnings (and may not even affect warnings on all OSs).

Also, are we actually getting any warnings from these libraries? Aka: Can we confirm this change actually has an effect?

**Review by weirddan455 — COMMENTED** [review]
It does build and seems to work on my Linux system. I have some concerns about letting an LLM play hack and slash with our build system very close to release.

**Comment by johnnovak** [comment]
> It does build and seems to work on my Linux system. I have some concerns about letting an LLM play hack and slash with our build system very close to release.

Oh, it's for 0.84! It goes into `main` only. We can keep merging stuff into `main`, then 0.83 fixes will go into `releases/0.83.x` and then forward ported.

Haha, it would only be hack and slash if I didn't review the changes and the CI loop wasn't passing 😏 Replying to your comments later. I also think Claude taught us many interesting techniques in these commits and is guiding us towards using CMake more idiomatically. I have certainly learnt a lot!

File: src/libs/glad/CMakeLists.txt (hunk 1 above)
Comment (johnnovak) [reply src/libs/glad/CMakeLists.txt]: Yeah, actually this was the single thing I was unsure about. I was 50/50 if it's worth the complicstion.

I think what it did was right, but given I'm planning to drop OpenGL anyway in favour of SDL3's GPU API this year for 0.84, this is a bit pointless. Dropping the entire commit.

File: CMakeLists.txt (hunk 3 above)
Comment (johnnovak) [reply CMakeLists.txt:372]: Nope, but it compiles, right? These are not functional changes, but build system changes. The change was about using the standard CMake provided lookup vs the old cruft, nothing else.

The commit message explains why the copy paste job from Ogre3D wasn't necessary.

I think it's easy to reason about this change; it's just lookup.

File: src/libs/glad/CMakeLists.txt (hunk 4 above)
Comment (johnnovak) [reply src/libs/glad/CMakeLists.txt]: No, the `SYSTEM` target was introduced with the explicit goal of silencing warnings in 3rd party code vendored into the source tree. Maybe the name is a bit confusing, but the KitWare guys were clear about the intent.

It's _not_ about literal system libs!

See the original issue ticket & discussion here:

https://gitlab.kitware.com/cmake/cmake/-/work_items/18040

> This will allow third-party code vendored within a project source tree to be treated as if imported as a system library.  Doing so will help suppress warnings in the third-party's header files when used from other first-party code in the same build.

Moreover:

> This will allow users to make use of more warnings without getting insane when using other projects like v8, rocksdb, s2geometry in their repository. Maintaining forks of those projects is hard and not all operating systems provide nice ready to install packages that will be found by CMake.

It's literally what we want to use it for. It's an elegant way to solve the problem, so a 🎖️ to Claude for bringing this into our attention 😎 

> Can we confirm this change actually has an effect?

It should if CMake is not broken 😄 
Will post a proof that it works later.

File: CMakeLists.txt (hunk 2 above)
Comment (johnnovak) [reply CMakeLists.txt]: @FeralChild64 can test this, yeah.

**Comment by johnnovak** [comment]
@weirddan455 You were absolutely right to push back on the `SYSTEM` stuff! 😎 

Seriously, though, I should've spent a bit more time testing this, no doubt. I think we're all learning how to use agents the best way to achieve solid results — the key point here was to put a lot of guardrails in and test everything locally, plus also test the negative cases (e.g. the warnings are back if we remove the new suppression mechanism).

I've basically spent the entire day on this (well, letting it work on it while cleaning the garage, then checking back every 30-60 minutes and give further instructions 😂), and I think the revised end result is extremely solid. Now we've _100% solved_ the warning problem with our vendored libs, and the solution is fully compliant with CMake recommendations and best practices (see reference links from the issue tracker).

Best of all, the problem is _also solved_ for header-only libs! These CMake guys have thought of everything 😎 

I've asked Claude to write a detailed account of what we've done exactly and how we have tested it; see below. We have done a _lot_ of testing in this round. We've also created a new `README.md` in `src/libs` that summarises our approach to suppressing warnings in vendored libs. Maybe start with reading that first.

As explained below, the two `TEST: ` commits are just there for local testing on macOS. I'll drop those from the final version once it's reviewed, but I thought I'd keep it for now; it might be helpful for re-testing stuff.

Another note: we've paid special attention _not_ to revert our patches to vendored libs that change the functionality. E.g., I remember applying some internal state defaulting (`= {}`) to TAL-Chorus and YM7128B_emu to get rid of some undefined random behaviour. Some internal structures were not initialised, and the random garbage sometimes created audible issues. None of those changes have been reverted, just the suppress warning stuff. Triple-checked that with Claude and also by reviewing the reverts.

PS: We've also fixed up the ManyMouse stuff. Told the thing to compare it against the Meson system, and make sure we link it the same way. But @FeralChild64 will still need to test it if he cares about it. In any case, that copy/pasted Ogre3D stuff was entirely unnecessary. I can't test it, so we've done some extra checks — see the next comment for details.

---

# Suppress compiler warnings from vendored third-party libraries — design and execution plan

## Problem

All code under `src/libs/` is vendored third-party code. Compiler warnings from these libraries obscure warnings from our own code. Previous attempts to suppress them were ad-hoc: pragma push/pop at include sites, per-file `-Wno-*` flags, source modifications (`[[maybe_unused]]`, `static_cast`, `(void)param`, etc.). These workarounds create maintenance burden — they must be re-applied every time a library is updated.

## Solution

Two complementary mechanisms cover all cases:

**1. SYSTEM includes** — tells the compiler to treat third-party headers as system headers, suppressing all warnings originating from them. For **header-only libraries**, this is a complete solution: their code only exists in headers, and every inclusion goes through `-isystem` via the INTERFACE target, so warnings are fully suppressed with no additional flags needed. For **compiled libraries**, SYSTEM suppresses warnings when project code includes their headers, but not when the library compiles its own sources (the compiler finds headers locally, bypassing `-isystem`).

**2. `-w` on compiled library targets** — suppresses all warnings during a library's own compilation. Adding `-w` (GCC/Clang) or `/w` (MSVC) via `target_compile_options(PRIVATE)` closes the gap that SYSTEM cannot reach. Only needed for compiled libraries, not header-only ones.

Together, these eliminate **_all warnings_** from third-party code without modifying vendored source files.

## Changes

### INTERFACE targets for header-only libraries

Header-only libraries previously had no CMake target — they relied on a blanket `src/libs` path in `project_headers`. Now each has an explicit INTERFACE target with SYSTEM includes, and consumers declare the dependency via `target_link_libraries`.

This is the officially recommended CMake pattern for header-only libraries: - [CMake Tutorial — In-Depth Library Concepts](https://cmake.org/cmake/help/latest/guide/tutorial/In-Depth%20CMake%20Library%20Concepts.html): *"The most common kind of interface library in C++ development is a
  header-only library."*
- [`cmake-buildsystem(7)`](https://cmake.org/cmake/help/latest/manual/cmake-buildsystem.7.html): *"A primary use-case for INTERFACE libraries is header-only libraries."*

New targets: `simpleini`, `json`, `base64`, `mverb`, `plmpeg`, `httplib`, `enet`, `clap`. `src/libs` removed from `project_headers` — third-party includes are now explicit per-target dependencies.

Also adds `PUBLIC ..` include directories to `loguru` and `manymouse` (compiled libraries that didn't expose their headers), and promotes `loguru` to PUBLIC linkage since its header is exposed through the project-wide `logging.h`.

### SYSTEM on all third-party include directories

Every library target (INTERFACE and compiled) now marks its include directories as SYSTEM. Consumers get `-isystem` instead of `-I`.

Also removes the redundant global `include_directories(SYSTEM .)` from `src/libs/CMakeLists.txt` — each library's own `target_include_directories` makes it unnecessary.

References:
- [`target_include_directories(SYSTEM)`](https://cmake.org/cmake/help/latest/command/target_include_directories.html): *"This may have effects such as suppressing warnings."*
- [CMake #18040](https://gitlab.kitware.com/cmake/cmake/-/issues/18040): SYSTEM target property for in-tree third-party code
- [Foonathan: Managing Compiler Warnings with CMake](https://www.foonathan.net/2018/10/cmake-warnings/)

### disable_warnings() for compiled library targets

A `disable_warnings()` helper function defined in `src/libs/CMakeLists.txt` applies `-w`/`/w` to a target. Each compiled vendored library calls it in its own `CMakeLists.txt`, keeping modules self-contained.

Applied to: decoders, esfmu, glad, loguru, manymouse, nuked, pdcurses, residfp, riffcpp, talchorus, whereami, ym7128bemu, zmbv.

### Include normalization

- Dropped `libs/` prefix from includes (`"libs/json/json.h"` → `"json/json.h"`) so they resolve through SYSTEM library targets instead of the non-SYSTEM `src` path
- Normalized angle-bracket includes of bundled headers to double quotes

### Removed workarounds

With SYSTEM + `-w` in place, all per-library warning hacks are removed:

**Consumption-site pragmas:**
- `shader_manager.cpp` — `-Weffc++` around SimpleIni include
- `misc_util.h` — `-Wdeprecated-*` around enet include

**CMake flag workarounds:**
- loguru — `-Wno-duplicated-branches`
- glad — `-Wno-cast-function-type-mismatch`

**Vendored source modifications** (reverted to match upstream):
- httplib — pragma shim wrapper in `http.h`
- ESFMu — pragma suppressions for `-Wlanguage-extension-token`, `-Wunused-variable`
- YM7128B_emu — `MAYBE_UNUSED` macro and 8 usages
- riffcpp — pragma blocks, `[[maybe_unused]]`, `static_cast` in 4 files
- ManyMouse — `__attribute__((unused))`, pragma for `-Wtautological-pointer-compare`
- TAL-Chorus — `= default` replacing `{}` destructors
- PDCurses — `[[maybe_unused]]` on 5 callback parameters

### Documentation

- Added `src/libs/README.md` documenting the warning-suppression approach, how to add new header-only and compiled libraries, and include conventions
- Added comment in root `CMakeLists.txt` explaining why warning flags are global (with vendored libs opting out via `disable_warnings()`)
- Fixed `LICENSE`: added glad, removed deleted ghc, fixed residfp typo (`GLP-2` → `GPL-2`), added Voodoo emulator (BSD-3-Clause), fixed stale paths for MAME and NE2000

## Testing

The branch includes TEST commits (artificial `-Wsign-compare` trigger in `SimpleIni.h`) that can be toggled to verify SYSTEM suppression on any platform.

**macOS results (Apple Clang, Xcode generator):**
- Without SYSTEM: warning fires from `SimpleIni.h`
- With SYSTEM: warning suppressed
- Without `-w` on compiled libs: warnings fire from reverted workarounds
- With `-w` restored: warnings suppressed
- Unit tests: 648 passed, 1 skipped, 0 failures
- Final clean build (313 TUs): zero warnings from vendored code

**Comment by johnnovak** [comment]
ManyMouse verification after a clean build:
 
image

File: CMakeLists.txt (hunk 3 above)
Comment (johnnovak) [reply CMakeLists.txt:372]: Double checked that IOKit is linked to the binary. See my last comment.

File: src/libs/glad/CMakeLists.txt (hunk 1 above)
Comment (johnnovak) [reply src/libs/glad/CMakeLists.txt]: Dropped.

File: CMakeLists.txt (hunk 2 above)
Comment (johnnovak) [reply CMakeLists.txt]: Reworked & retested, see my comments on the main thread.

File: src/libs/glad/CMakeLists.txt (hunk 4 above)
Comment (johnnovak) [reply src/libs/glad/CMakeLists.txt]: See my comments on the main thread.

File: src/libs/manymouse/CMakeLists.txt [hunk 5]
```diff
@@ -12,8 +12,12 @@ if(C_MANYMOUSE)
     x11_xinput2.c
   )
 
-  target_link_libraries(dosbox PRIVATE
-    $<IF:$<BOOL:${SUPPORT_XINPUT2}>,${X11_Xinput_LIB},>
-  )
+  if (APPLE)
+    target_link_libraries(libmanymouse PRIVATE "-framework IOKit")
+  endif()
```
Comment (weirddan455) [inline src/libs/manymouse/CMakeLists.txt:18]: There's already code relating to IOKit in the top level CMakeLists.txt.

```
elseif (DOSBOX_PLATFORM_MACOS)
  find_library(IOKit_LIBRARY IOKit)
  if (NOT IOKit_LIBRARY)
    message(WARNING "ManyMouse requires IOKit")
  else()
    set(C_MANYMOUSE ON)
  endif()
```

I believe the more idiomatic way to link is to use the variable we told `find_library` to store the result in like `target_link_libraries(libmanymouse PRIVATE ${IOKit_LIBRARY}")`

This is also what we're doing on the next line for the X11 library.

File: src/gui/CMakeLists.txt [hunk 6]
```diff
@@ -16,7 +16,7 @@ target_sources(dosboxcommon PRIVATE
   render/shader_pipeline.cpp
 )
 
-target_link_libraries(dosboxcommon PRIVATE OpenGL::GL)
+target_link_libraries(dosboxcommon PRIVATE OpenGL::GL simpleini)
```
Comment (weirddan455) [inline src/gui/CMakeLists.txt:19]: This commit is a whole bunch of extra code that looks like it's tracking dependencies of header-only libraries but it doesn't actually _do_ anything. For example, I can remove `simpleini` here and everything still works.

Take a look at `compile_commands.json` and see what it's actually passing to the compiler when it builds `shader_manager.cpp` (callsite of simpleini):

```
{
  "directory": "/home/daniel/code/dosbox-staging/build/debug-linux",
  "command": "/usr/bin/clang++ -I/home/daniel/code/dosbox-staging/src -I/home/daniel/code/dosbox-staging/build/de
bug-linux/include -isystem /home/daniel/code/dosbox-staging/src/misc/../libs/include -isystem /home/daniel/code/d
osbox-staging/src/libs/clap/.. -isystem /home/daniel/code/dosbox-staging/src/libs/mverb/.. -isystem /home/daniel/
code/dosbox-staging/src/libs/zmbv/.. -isystem /home/daniel/code/dosbox-staging/src/libs/simde/.. -isystem /usr/in
clude/SDL2 -isystem /home/daniel/code/dosbox-staging/src/libs/simpleini/.. -isystem /home/daniel/code/dosbox-stag
ing/src/libs/glad/debug/include -isystem /home/daniel/code/dosbox-staging/src/libs/enet/.. -isystem /home/daniel/
code/dosbox-staging/src/libs/ESFMu/.. -isystem /home/daniel/code/dosbox-staging/src/libs/residfp/.. -isystem /hom
e/daniel/code/dosbox-staging/src/libs/YM7128B_emu/. -isystem /home/daniel/code/dosbox-staging/src/libs/riffcpp/. 
-isystem /home/daniel/code/dosbox-staging/src/libs/tal-chorus/.. -isystem /home/daniel/code/dosbox-staging/src/li
bs/nuked/.. -isystem /home/daniel/code/dosbox-staging/src/libs/PL_MPEG/.. -isystem /home/daniel/code/dosbox-stagi
ng/src/libs/manymouse/.. -isystem /home/daniel/code/dosbox-staging/src/libs/whereami/. -isystem /home/daniel/code
/dosbox-staging/src/libs/base64/.. -isystem /home/daniel/code/dosbox-staging/src/libs/http/.. -isystem /home/dani
el/code/dosbox-staging/src/libs/json/.. -isystem /home/daniel/code/dosbox-staging/src/libs/loguru/.. -isystem /ho
me/daniel/code/dosbox-staging/src/libs/decoders/.. -g -std=c++20 -fcolor-diagnostics -Wall -Wextra -Wdeprecated -
Wno-unknown-warning-option -Wno-unknown-pragmas -Wno-conversion -Wno-narrowing -Wdisabled-optimization -Wduplicat
ed-branches -Wtrampolines -Wvla -Wlogical-op -Wduplicated-cond -Wpointer-arith -Wredundant-decls -Wfloat-conversi
on -Weffc++ -Wextra-semi -Wimplicit-fallthrough -Winvalid-utf8 -Wtrivial-auto-var-init -o CMakeFiles/dosboxcommon
.dir/src/gui/render/shader_manager.cpp.o -c /home/daniel/code/dosbox-staging/src/gui/render/shader_manager.cpp",
  "file": "/home/daniel/code/dosbox-staging/src/gui/render/shader_manager.cpp",
  "output": "/home/daniel/code/dosbox-staging/build/debug-linux/CMakeFiles/dosboxcommon.dir/src/gui/render/shader_manager.cpp.o"
},
```

Every one of these libs is including with `-isystem` a relative path back to its parent in `src/libs`. It's just including that single directory multiple times.

It works but it's just adding a bunch of complexity for no real benefit. You would get the same result with a single line of code `target_include_directories(dosboxcommon SYSTEM PUBLIC src/libs)` rather than creating interfaces for all these header only libraries that just aren't doing anything.

File: src/libs/zmbv/CMakeLists.txt [hunk 7]
```diff
@@ -1,3 +1,5 @@
 add_library(zmbv STATIC zmbv.cpp)
 target_include_directories(zmbv SYSTEM PUBLIC ..)
 target_link_libraries(zmbv PRIVATE project_headers ${ZMBV_ZLIB_TARGET})
+
+disable_warnings(zmbv)
```
Comment (weirddan455) [inline src/libs/zmbv/CMakeLists.txt:5]: I think we do want warnings for ZMBV and maybe it shouldn't even be in `src/libs`. There is no upstream project here. As far as I know ZMBV is a file format developed by the DOSBox devs so the code is essentially ours to own.

**Review by weirddan455 — COMMENTED** [review]
> @weirddan455 You were absolutely right to push back on the SYSTEM stuff! 😎

That's the most AI sounding line I've ever seen 😆 

It turns out you were absolutely right about `-isystem` hiding warnings for header-only libraries. This PR does just that for the GCC warnings I was getting in SimpleIni in #4860

See code comments below. I think some of this can be simplified but I like the approach. It's a lot cleaner than sticking those `#pragma`'s everywhere.

As long as we're sticking to upstream on these vendored libraries I think it's fine to hide all warnings. We need to be careful about libraries we're making changes to because we would want to be warned if we do something dumb in a PR.

**Comment by FeralChild64** [comment]
>  But @FeralChild64 will still need to test it if he cares about it

@johnnovak Frankly, I have lost the motivation. I was hoping I would be able to improve the build process to be acceptable for me (fully offline build again), but:
- I have read about the MkDocs 1.x being unmaintained and that there is some ongoing drama among the former developers.
- On the Zensical FAQ I have found the information that MkDocs is now becoming a "supply chain risk".
- I have witnessed Gentoo kicking all the MkDocs packages out of the main repo due to various problems (I've seen the comments in the ebuilds...), high maintenance effort needed, and (according to Gentoo team) almost no open-source software using it.
- I have seen the size of dependencies needed by the MkDocs packages used.
- I have read your post about the amount of customizations made for DOSBox Staging.

I don't want to work with something like this (at least not in my spare time; professional work I'm payed to do full time is a different story), I don't want to promote it, I'm not even sure if it's a good idea to install this on my PC, and I definitely don't have enough time and energy.

Plus, even if I find a way to make the manual/guide build process sane in my opinion, I have no guarantee that in a few months something else won't gets merged that breaks the offline compilation again.

I can answer questions regarding mouse/keyboard/locale (if you have any), maybe review changes to these parts, but that's mostly it - I've stopped closely following what's happening in the project, I'm not really interested in testing builds not coming from my Linux distro package manager.

**Comment by johnnovak** [comment]
> I was hoping I would be able to improve the build process to be acceptable for me (fully offline build again), but:

By default we don't build the doco, you need to pass the `OPT_DOCUMENTATION` CMake option for that.

> mkdocs drama, customisations

ProperDocs is a direct fork of MkDocs that gets updates, and MaterialX is a direct fork of MkDocs Material:

https://github.com/jaywhj/mkdocs-materialx

They're literal drop-in replacements; all customisations and all hooks just work.

At work, people have made the switch already, literally a few lines' worth of change. I was in no particular hurry because MkDocs will just keep working even in its current state forever, just no new features. That's the beauty of static site generators -- we only really care about the output.

I could migrate in an hour probably.
With Zensical, we'll see where things will go, so it's not a good time to change to that yet (or ever).

> I don't want to work with something like this (at least not in my spare time; professional work I'm payed to do full time is a different story), I don't want to promote it, I'm not even sure if it's a good idea to install this on my PC, and I definitely don't have enough time and energy.

You're one very finicky individual, did anyone tell you that yet? 😂 Surely, you install Python stuff on your machine with `pip` all the time. So you can install all the current toolchain just like the CI job does with a single `pip` command.

I can also do the ProperDocs + MaterialX transition earlier if that's the _only_ thing throwing the spanner in the works for you. One hour job for me, probably, rather trivial. 

> I can answer questions regarding mouse/keyboard/locale (if you have any), maybe review changes to these parts, but that's mostly it - I've stopped closely following what's happening in the project,

Your life, your time, of course 😄 

> I'm not really interested in testing builds not coming from my Linux distro package manager.

Yeah, so as I said, if you want to test this PR locally for the ManyMouse stuff, you can. Doco building is opt-in, that should be clear to you by now.

In fact, you could easily do the local documentation update workflow as well -- you'd be "exposed" to the exact same "security risk" by building the doco and opening it in your browser for local viewing as when visiting our website. Or another MkDocs generated website. Or any random website on the Internet, really 🤷🏻 You are really making a big fuss out of nothing, but whatever man 🤷🏻‍♂️ 

FYI, we'll probably drop the ManyMouse stuff from the project at some point if you'll stop contributing as it's a high complexity, hard to test, and extremely niche feature.

File: src/gui/CMakeLists.txt (hunk 6 above)
Comment (johnnovak) [reply src/gui/CMakeLists.txt:19]: I think I'd like to keep it like this as using interfaces for header-only libs is the recommended way in the official CMake guide.

I don't see this as noise, but rather a welcome feature. I like it that we are explicit about the dependencies. On the contrary, I see _not_ listing header-only libs as deps as a weird anomaly. They _are_ dependencies, they are vendored, so why should they be "silent", unlisted deps?

I get it that it might work without this, but I see that as an irregularity and special-casing header-only libs for just saving a few characters.

I agree with the recommended CMake approach here -- treat them just like regular libs, just via the INTERFACE type. As I see it, that's the most consistent approach. View it as documentation in the CMake configs if nothing else. It's about having a sound and complete dependency graph; the rest is really just implementation detail.

File: src/libs/zmbv/CMakeLists.txt (hunk 7 above)
Comment (johnnovak) [reply src/libs/zmbv/CMakeLists.txt:5]: Yeah, good point. I'll move it into our main source tree. I think the historical reason for this was to be able to build a standalone zmbv lib for integrating it into other apps. But that be achieved in the main source tree, too -- this is not a vendored lib.

**Comment by johnnovak** [comment]
> As long as we're sticking to upstream on these vendored libraries I think it's fine to hide all warnings. We need to be careful about libraries we're making changes to because we would want to be warned if we do something dumb in a PR

That's a good point. I should add a note to the README with a warning about this, saying we should temporarily comment out the warning suppression while working on patches that won't be upstreamed. Then cross fingers that peiple will actually read the doco 😆

**Comment by johnnovak** [comment]
@weirddan455 I think i'll move ZMBV in a followup PR. Are you happy with this then?

File: src/libs/manymouse/CMakeLists.txt (hunk 5 above)
Comment (johnnovak) [reply src/libs/manymouse/CMakeLists.txt:18]: Yeah good point, addressed.

**Comment by weirddan455** [comment]
> @weirddan455 I think i'll move ZMBV in a followup PR. Are you happy with this then?

The number of commits doubles every time I look at this PR...

**Comment by johnnovak** [comment]
> > @weirddan455 I think i'll move ZMBV in a followup PR. Are you happy with this then?
> 
> The number of commits doubles every time I look at this PR...

Uh oh, my bad. I also have a WIP PVS cleanup PR that somehow got mixed into this. I'm working on shit on 3 machines in parallel 😅 Will fix it once I get home from the gym 😎

**Comment by johnnovak** [comment]
@weirddan455 Sanity has been restored. The management apologises for any inconvenience caused.

File: src/gui/CMakeLists.txt (hunk 6 above)
Comment (weirddan455) [reply src/gui/CMakeLists.txt:19]: The point of a header-only library is that you don't have to mess with a build system. You `#include` and it just works. That's it. Why over-complicate things?

> View it as documentation in the CMake configs if nothing else.

That's literally all it's doing and if you view it as documentation it's very misleading. You call `target_link_libraries` on these header-only libraries and there is absolutely no linking taking place.

I'm not going to argue this any further. If you want to go this route, sure. It works at least.

Comment (kklobe) [reply src/gui/CMakeLists.txt:19]: Your approach seems consistent with recommended practices @johnnovak.

From [Professional CMake](https://crascit.com/professional-cmake/):

File: website/docs/0.83/manual/using-dosbox-staging/shell.md [hunk 8]
```diff
@@ -54,7 +54,7 @@ See [DOS commands](commands.md) for the complete command reference.
 The shell supports full command-line editing. You can move the cursor, insert
 and delete characters, and recall previous commands --- all features that
 required a separate `DOSKEY` utility on real MS-DOS systems (shipped with
-MS-DOS 5.0 in 1991), but are built into DOSBox Staging's shell.
+MS-DOS 5.0 in 1991), but are built into DOSBox Staging's DOS shell.
```
Comment (weirddan455) [inline website/docs/0.83/manual/using-dosbox-staging/shell.md:57]: Remove the "but".

Is this commit even supposed to go with this PR?

**Review by weirddan455 — APPROVED** [review]
Glanced through this all again and I'm happy enough with it. Merge when ready.

**Comment by FeralChild64** [comment]
> By default we don't build the doco, you need to pass the `OPT_DOCUMENTATION` CMake option for that.

But this documentation is now a part of the package, and apparently all the contributors are required to update it and this is non-negotiable.
From my side, I'm not interested in software which needs to fetch something from the Internet during the build process, and this is also non-negotiable.

> I could migrate in an hour probably.

I might have a look once again when some modern fork is used - no promises, though.

> Surely, you install Python stuff on your machine with pip all the time.

What for? The _Portage_ tree contains a _dev-python_ directory, probably all the commonly used packages are there - and all the needed files are stored on multiple _Gentoo_ mirrors all over the world, their checksums are verified, the authenticity and integrity of the Gentoo `Portage` tree is verified when updating using _OpenPGP_ signatures, Gentoo automatically rebuilds packages if needed after _Python_ upgrade, it tracks dependencies to non-Python packages, etc.

> Yeah, so as I said, if you want to test this PR locally for the ManyMouse stuff, you can. Doco building is opt-in, that should be clear to you by now.

I'm not following closely what's happening here, so I'm not sure. OK, I've built the emulator from this PR and it looks like the _ManyMouse_ is still working, at least on Linux.

> In fact, you could easily do the local documentation update workflow as well -- you'd be "exposed" to the exact same "security risk" by building the doco and opening it in your browser for local viewing as when visiting our website. Or another MkDocs generated website. Or any random website on the Internet, really

Nope. Modern web browsers are quite good at protecting the user against malicious web scripts, while Python virtual environment does not even try to be a secure container. Plus, I have other concerns - like future availability of the web resources (few times in my life I had to compile a piece of software which was like 40 years old...), or the effort needed to package everything for the Linux distros.

> FYI, we'll probably drop the ManyMouse stuff from the project at some point if you'll stop contributing as it's a high complexity, hard to test, and extremely niche feature.

I was intending to drop it soon after migration to SDL 3 - it's updated mouse API allows to implement the multi-mouse support in a much cleaner way; apparently _Amiberry_ is already using it.

File: src/gui/CMakeLists.txt (hunk 6 above)
Comment (johnnovak) [reply src/gui/CMakeLists.txt:19]: Yeah, we'll use the recommended approach. As per the above quote, this has other benefits, too, apart from having a complete build graph and consistency.

I'd also argue the original intent and selling point of header-only libs was not having to use a build system _at all_. But we're using one, so the advantage of not having to use a build system does not apply. Therefore, the official CMake recommended approach it is. I always prefer correctness to shortcuts, especially if the shortcuts introduce irregularities.

File: website/docs/0.83/manual/using-dosbox-staging/shell.md (hunk 8 above)
Comment (johnnovak) [reply website/docs/0.83/manual/using-dosbox-staging/shell.md:57]: Nah, removing. Removing the "but" too 😆 I keep pushing tons of doco changes directly to `release/0.83.x` the forward pot to `main`.

You guys will get to review the entire doco before 0.83 final; until they I work on it in my own chaotic non-linear ways 😎  After that, we'll have a solid base to make incremental changes as a sane, linear process via PRs.

**Comment by johnnovak** [comment]
> OK, I've built the emulator from this PR and it looks like the _ManyMouse_ is still working, at least on Linux.

Thanks for confirming.

> I was intending to drop it soon after migration to SDL 3 - it's updated mouse API allows to implement the multi-mouse support in a much cleaner way; apparently _Amiberry_ is already using it.

Cool, you can do that then if you want after the SDL3 branch gets merged, or we'll ask you for advice and we'll do it.

> From my side, I'm not interested in software which needs to fetch something from the Internet during the build process, and this is also non-negotiable

You can still work on the project without downloading anything.

1. Default build doesn't download anything.
2. Locally work on the doco with live preview, `cd website` and `mkdocs serve`. Local preview viewable on localhost, it's a live preview.
3. Make .md changes, check preview, commit, push, done. 

Easy. No downloads. All described in the doco readme.

If you're super obsessed with those deps, you can host them on our GH pages website and change the HTML template in one place. There is always a way -- unless one makes up his mind and hyperfocuses on why he _cannot_ do it 😎

File: website/docs/0.83/manual/using-dosbox-staging/shell.md (hunk 8 above)
Comment (johnnovak) [reply website/docs/0.83/manual/using-dosbox-staging/shell.md:57]: Rogue commit dropped.
